File: src/DesignPatterns/DIY/Template/Template.java
package DesignPatterns.DIY.Template;

public class Template {
    public static void main(String[] args) {
        MainCharacter mainCharacter = new MainCharacter();
        mainCharacter.createCharacter();

    }
}

abstract class Character{

    private DataCenter dataCenter;

    public Character(DataCenter dataCenter) {
        this.dataCenter = dataCenter;
    }
    public Character() {
        dataCenter = new DataCenter();
    }

    public void createCharacter(){
        dataCenter.server();
        characterName();
        selectRace();
        selectClass();
        chooseStartingCity();
    }
    protected abstract void characterName();
    protected abstract void selectRace();
    protected abstract void selectClass();
    protected abstract void chooseStartingCity();

}

class DataCenter {
    public void server(){
        System.out.println("Server: Atomos");
    }
}

class MainCharacter extends Character {
    
    @Override
    protected void characterName() {
        System.out.println("Name: Welt Yang");
    }

    @Override
    protected void selectRace() {
        System.out.println("Race: Catgirl");
    }

    @Override
    protected void selectClass() {
        System.out.println("Class: Dragoon");
    }

    @Override
    protected void chooseStartingCity() {
        System.out.println("Starting City: Gridania");
    }

}

File: src/DIY/Template/Template.ts
class DataCenter {
    server(): void {
      console.log("Server: Atomos");
    }
  }
  
  abstract class Character {
    private dataCenter: DataCenter;
  
    constructor(dataCenter?: DataCenter) {
      this.dataCenter = dataCenter ?? new DataCenter();
    }
  
    createCharacter(): void {
      this.dataCenter.server();
      this.characterName();
      this.selectRace();
      this.selectClass();
      this.chooseStartingCity();
    }
  
    protected abstract characterName(): void;
    protected abstract selectRace(): void;
    protected abstract selectClass(): void;
    protected abstract chooseStartingCity(): void;
  }
  
  class MainCharacter extends Character {
    protected characterName(): void {
      console.log("Name: Welt Yang");
    }
  
    protected selectRace(): void {
      console.log("Race: Catgirl");
    }
  
    protected selectClass(): void {
      console.log("Class: Dragoon");
    }
  
    protected chooseStartingCity(): void {
      console.log("Starting City: Gridania");
    }
  }
  
  (function main(): void {
    const mainCharacter = new MainCharacter();
    mainCharacter.createCharacter();
  })();
  